crates/ruff/src/message/diff.rs

```diff
@@ -0,0 +1,194 @@
+use crate::message::Message;
+use colored::{Color, ColoredString, Colorize, Styles};
+use ruff_diagnostics::Fix;
+use ruff_python_ast::source_code::{OneIndexed, SourceCode};
+use ruff_python_ast::types::Range;
+use ruff_text_size::{TextRange, TextSize};
+use similar::{ChangeTag, TextDiff};
+use std::fmt::{Display, Formatter};
+use std::num::NonZeroUsize;
+
+/// Renders a diff that shows the code fixes.
+///
+/// The implementation isn't fully fledged out and only used by tests. Before using in production, try
+/// * Improve layout
+/// * Replace tabs with spaces for a consistent experience across terminals
+/// * Replace zero-width whitespaces
+/// * Print a simpler diff if only a single line has changed
+/// * Compute the diff from the [`Edit`] because diff calculation is expensive.
+pub(super) struct Diff<'a> {
+    fix: &'a Fix,
+    source_code: SourceCode<'a, 'a>,
+}
+
+impl<'a> Diff<'a> {
+    pub fn from_message(message: &'a Message) -> Option<Diff> {
```

Comment:
Should this be `try_from_message`? Or is `try_from_*` only used for `Result`?

Reply:
Not sure. `NonZeroUsize` returns `Option` from `new`. My reasoning is that it isn't failing, it's just that not all `Message`s can print a `diff`.